File: spring-web-java/src/main/java/spring/web/java/domain/model/entity/Order.java
package spring.web.java.domain.model.entity;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.ConstraintMode;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.ForeignKey;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import spring.web.java.domain.model.enums.DeliveryStatus;
import spring.web.java.domain.model.enums.OrderStatus;
import spring.web.java.presentation.exception.OrderCancelNotAllowedException;

@Entity
@Getter
@Table(name = "orders")
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Order extends Base {

	@Id
	@GeneratedValue
	@Column(name = "order_id")
	private Long id;

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "member_id", foreignKey = @ForeignKey(ConstraintMode.NO_CONSTRAINT))
	private Member member;

	@OneToMany(mappedBy = "order", cascade = CascadeType.ALL)
	private List<OrderItem> orderItems = new ArrayList<>();

	@OneToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
	@JoinColumn(name = "delivery_id", foreignKey = @ForeignKey(ConstraintMode.NO_CONSTRAINT))
	private Delivery delivery;

	private Instant orderedAt;

	@Enumerated(EnumType.STRING)
	private OrderStatus status;

	public List<OrderItem> getOrderItems() {
		return Collections.unmodifiableList(orderItems);
	}

	public static Order create(Member member, Delivery delivery, List<OrderItem> orderItems) {
		Order order = new Order();

		order.member = member;
		order.associateDelivery(delivery);
		orderItems.forEach(order::associateItem);
		order.status = OrderStatus.ORDER;
		order.orderedAt = Instant.now();

		return order;
	}

	public void associateItem(OrderItem orderItem) {
		orderItems.add(orderItem);
		orderItem.setOrder(this);
	}

	public void associateDelivery(Delivery delivery) {
		this.delivery = delivery;
		delivery.setOrder(this);
	}

	public void cancel() {
		if (delivery.getStatus() == DeliveryStatus.COMP) {
			throw new OrderCancelNotAllowedException(this.id);
		}

		status = OrderStatus.CANCEL;
		orderItems.forEach(OrderItem::cancel);
		delivery.cancel();
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/domain/model/entity/Order.kt
package spring.web.kotlin.domain.model.entity

import jakarta.persistence.*
import spring.web.kotlin.domain.model.enums.OrderStatus
import spring.web.kotlin.presentation.exception.OrderCancelNotAllowedException
import java.time.Instant

@Entity
@Table(name = "orders")
class Order protected constructor(
    val orderedAt: Instant,
    status: OrderStatus,
    member: Member
) : Base() {
    @Id
    @GeneratedValue
    @Column(name = "order_id")
    var id: Long? = null
        protected set

    @Enumerated(EnumType.STRING)
    var status: OrderStatus = status
        protected set

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id", foreignKey = ForeignKey(ConstraintMode.NO_CONSTRAINT))
    var member: Member = member
        protected set

    @OneToMany(mappedBy = "order", cascade = [CascadeType.ALL])
    private val _orderItems: MutableList<OrderItem> = mutableListOf()

    @get:Transient
    val orderItems: List<OrderItem>
        get() = _orderItems.toList()

    companion object {
        fun create(member: Member, orderItems: List<OrderItem>) =
            Order(
                orderedAt = Instant.now(),
                status = OrderStatus.ORDER,
                member = member,
            ).apply { orderItems.forEach { associateItem(it) } }
    }

    fun associateItem(orderItem: OrderItem) {
        _orderItems.add(orderItem)
        orderItem.associateOrder(this)
    }

    fun cancel() {
        val id = requireNotNull(this.id)

        if (status == OrderStatus.CONFIRM) {
            throw OrderCancelNotAllowedException(id)
        }

        status = OrderStatus.CANCEL
        orderItems.forEach { it.cancel() }
    }
}
